pkg/services/opsgenie/opsgenie_config.go

```diff
@@ -0,0 +1,121 @@
+	Port        uint16            `desc:"The OpsGenie API port." default:"443"`
+	Alias       string            `key:"alias" desc:"Client-defined identifier of the alert" optional:"true"`
+	Description string            `key:"description" desc:"Description field of the alert" optional:"true"`
+	Responders  []Entity          `key:"responders" desc:"Teams, users, escalations and schedules that the alert will be routed to send notifications" optional:"true"`
+	VisibleTo   []Entity          `key:"visibleTo" desc:"Teams and users that the alert will become visible to without sending any notification" optional:"true"`
+	Actions     []string          `key:"actions" desc:"Custom actions that will be available for the alert" optional:"true"`
+	Tags        []string          `key:"tags" desc:"Tags of the alert" optional:"true"`
+	Details     map[string]string `key:"details" desc:"Map of key-value pairs to use as custom properties of the alert" optional:"true"`
+	Entity      string            `key:"entity" desc:"Entity field of the alert that is generally used to specify which domain the Source field of the alert" optional:"true"`
+	Source      string            `key:"source" desc:"Source field of the alert" optional:"true"`
+	Priority    string            `key:"priority" desc:"Priority level of the alert. Possible values are P1, P2, P3, P4 and P5" optional:"true"`
+	Note        string            `key:"note" desc:"Additional note that will be added while creating the alert" optional:"true"`
+	User        string            `key:"user" desc:"Display name of the request owner" optional:"true"`
+}
+
+func (config Config) Enums() map[string]types.EnumFormatter {
+	return map[string]types.EnumFormatter{}
+}
+
+// Public version of GetURL that creates a new PropKeyResolver when accessed from outside the package
+func (config *Config) GetURL() *url.URL {
+	resolver := format.NewPropKeyResolver(config)
+	return config.getURL(&resolver)
+}
+
+// Private version of GetURL that can use an instance of PropKeyResolver instead of rebuilding it's model from reflection
+func (config *Config) getURL(resolver types.ConfigQueryResolver) *url.URL {
+	host := ""
+	if config.Port > 0 {
+		host = fmt.Sprintf("%s:%d", config.Host, config.Port)
+	} else {
+		host = config.Host
+	}
+
+	// Serialize Shoutrrr standard data types
+	rawQuery := format.BuildQuery(resolver)
+
+	// Serialize OpsGenie specific data types
+	separator := "&"
+	if rawQuery == "" {
+		separator = ""
+	}
+
+	if len(config.Responders) > 0 {
+		responders, _ := serializeEntities(config.Responders)
+		rawQuery = rawQuery + separator + "responders=" + responders
+		separator = "&"
+	}
+	if len(config.VisibleTo) > 0 {
+		visibleTo, _ := serializeEntities(config.VisibleTo)
+		rawQuery = rawQuery + separator + "visibleTo=" + visibleTo
+	}
+
+	result := &url.URL{
+		Host:     host,
+		Path:     fmt.Sprintf("/%s", config.ApiKey),
+		Scheme:   Scheme,
+		RawQuery: rawQuery,
+	}
```

Comment:
This could be made a little bit simpler by using the `url.Values` for manipulating the query:
```suggestion
	// Serialize Shoutrrr standard data types
	query, _ := url.ParseQuery(format.BuildQuery(resolver))

	// Serialize OpsGenie specific data types
	if len(config.Responders) > 0 {
		responders, _ := serializeEntities(config.Responders)
		query.Set("responders", responders)
	}
	if len(config.VisibleTo) > 0 {
		visibleTo, _ := serializeEntities(config.VisibleTo)
		query.Set("visibleTo", visibleTo)
	}

	result := &url.URL{
		Host:     host,
		Path:     fmt.Sprintf("/%s", config.ApiKey),
		Scheme:   Scheme,
		RawQuery: query.Encode(),
	}
```

Reply:
I tried that. The problem is that `query.Encode()` encodes commas as well so something like `tags=t1,t2` becomes tags=t1%2ct2`

Reply:
Ok fixed